### ISSUE TITLE
More docs about how to update/maintain an Elgg site with composer URGENTLY needed

### ISSUE THREAD

**Issue by iionly** [Elgg/Elgg#10303]
The docs have instructions of how to install Elgg with composer. But that's all.
- There's no info about how to update Elgg to a newer version with composer.
- There's no info about how to install 3rd party plugins.
- There's no troubleshooting info.

I got a PR for the Event Calendar to add instructions of how to add it to an installation via composer. Now, I don't use composer so far - apart from testing an installation via composer. Before just merging the PR blindly I wanted to test the instructions of course and I came across some problems then:
- A simple "composer require iionly/event_calendar" did not work. I had to run "composer update" first to update the dependencies of Elgg.
- The "composer update" caused jquery-ui to be updated to version 1.12.1 and that is a BC-breaking change just resolved in Elgg 2.2.1 (the Elgg install is still on 2.0.4).
- The idea of fixing the jquery-ui BC-breaking issue by just requiring a newer version of elgg/elgg resulted in the problem that jquery-ui version 1.11.\* was requested but "no matching package was found".

Now I guess I have a broken Elgg installation without any idea of how to fix it. In such a situation it would be REALLY helpful to have some info in the Elgg docs about how to fix the problem (or even better to tell me how to do it the right way in the first place without breaking anything at all).

I think it's really urgently necessary to provide sufficient info in the docs for common tasks - and not just saying that all the info is to be found somewhere on the Net. Most information about composer is very general and directed to developers who want to manage the dependencies of their projects and not to "end-users" who just want to set up / maintain a website / Elgg site.

If any useful information about how to manage 3rd party plugins with composer in general would be available, it would also be much easier (or possible at all) to mention the composer installation option in a plugin's readme. It would be enough to link to the packagist repo and to link to the general info in the Elgg docs without the risk to get asked about how to fix a broken Elgg installation caused by trying to add a plugin via composer.

**Comment by mrclay** [comment]
`composer.lock` defines what `composer install` does, so revert changes to that file and reinstall via composer. When using `composer require` always specify a version of the dependency so that composer won't do an update to pull in everything new. The composer docs don't mention this!

**Comment by iionly** [comment]
The initial "problem" was that a "composer require iionly/event_calendar ~2.0" did not work (even with the version I want specified). Only after a "composer update" it worked. I don't know why this was necessary as there are no specific dependencies required by event_calendar apart from elgg/elgg.

But the "composer update" did update all dependencies including updating jquery-ui to version 1.12.1 that I knew would cause a problem. composer.lock did exist at that point and I hadn't made any changes to it. How would I know which dependencies would be updated and which updates would be necessary?

The idea of updating Elgg from 2.0.4 (which is still the version used in the elgg-starter project) to 2.2.1 to "fix" the issue with jquery-ui was just kind of a test as I wanted to see if this would fix the jquery-ui version issue by downgrading automatically. "composer require elgg/elgg 2.2.1" did not work though as composer did fail to handle the jquery-ui dependency. composer tried to install "1.11.*" of jquery-ui and found versions 1.11.0 - 1.11.4 all to match but then gave up.

At that point I deleted composer.lock, made a "composer require bower-asset/jquery-ui 1.11.4" followed by an "composer update" and then I had Elgg version 2.2.1 finally.

So, I could fix my problem with some trial and error (was a test in the first place anyway). But I described my problem just as an example that there might be common issues when managing an Elgg site with composer and there's currently no info in the docs about that. If people get nudged to use composer for installing Elgg and also 3rd party plugins it might be of help to have some additional info in the docs for sure.

**Comment by beck24** [comment]
I think @iionly might have a point here, we're moving to composer for managing everything, and there was talk about possibly ending the zip builds in favour of it.  Having some common task examples for it would probably be a good first step towards that.

**Comment by hypeJunction** [comment]
Someone just needs to do it.

**Comment by iionly** [comment]
If I wouldn't be in need for these kind of docs myself I would give it a try.

But writing down such info might have some benefits in the future, i.e. less stupid questions if people can read it on their own - or at least you could give them a RTFM if they still ask. ;-)

**Comment by mrclay** [comment]
Sending some docs PRs to composer/composer would help as well once we figure this out.

**Comment by iionly** [comment]
I don't care anymore about that.

**Comment by jdalsem** [comment]
we care

**Comment by hypeJunction** [comment]
Might come in handy
https://gist.github.com/hypeJunction/b1f89b829dfa669974cc465a46d3ab07
